Feature: Filtering on Scalar fields

sMteX opened the pull request:
## EDIT: See latest comment for current info on the MR

For `DateTime` Prisma fields, there is a `DateTimeFilter` and `DateTimeNullableFilter` input type generated, in the form of 
```gql
input DateTimeFilter {
  equals: DateTime
  in: [DateTime]
  notIn: [DateTime]
  lt: DateTime
  lte: DateTime
  gt: DateTime
  gte: DateTime
  not: NestedDateTimeFilter
}
```

Before this merge request, if you tried to use `DateTime` field in react-admin's `<Filter>`, e.g.

```jsx
<DateInput source="effectiveFrom" />
```

the data provider would not recognize it's a `DateTimeFilter` and produces a query like this:

```js
where: { effectiveFrom: { id: { equals: '2020-12-30' } } }
```

which would turn into a 400 Bad Request for the query.

This merge request implements basic filtering on `DateTime` fields by using an "artificial" field name (inspiration taken from React Admin's [demo](https://github.com/marmelab/react-admin/blob/master/examples/demo/src/orders/OrderList.tsx#L48)). Now, you can filter DateTime fields by equal comparison, or with operations >, >=, <, <= in the following way (taking a field `effectiveFrom` as an example):

- using `source="effectiveFrom"` without any suffix generates 
    `where: { effectiveFrom: { equals: '2020-12-30T00:00:00.000Z' } }`
- using `source="effectiveFrom_lt"` uses lesser than comparison and generates 
    `where: { effectiveFrom: { lt: '2020-12-30T00:00:00.000Z' } }`
- similarly for `lte`, `gt`, `gte`

The technical limitation of this approach is that users can't use `_` in **any** field names (in which case, a different separator, maybe `:` would be a safer choice for users with snake_case field names). 

I also included several tests (and others passed as well so I think this change shouldn't break anything else), but generating the types (via `yarn generate:test:nexus`) failed for me so I'm not sure if I didn't make any mistakes there.

TODO: If you would want, I could try to implement the rest of the filter fields (`in`, `notIn`, `not`) as well, but for our use case this wasn't necessary.

macrozone commented:
@sMteX thats a nice approach and tests look also good!

> The technical limitation of this approach is that users can't use _ in any field names (in which case, a different separator, maybe : would be a safer choice for users with snake_case field names).

You can check whether the Graphql input type actually supports your query. The information is already there from introspection.

So you would only assume the special meaning of `_` if it results in a valid query

sMteX commented:
@macrozone 

Actually I looked more closely at the regular expression, and if we change it slightly to 

```js
/^(.+?)(_(gt|gte|lt|lte))?$/
```

so the first group now accepts all characters but _lazily_, then even strings like "effective_from_gt" get correctly parsed into "effective_from" and "gt".

Do you think that would solve the issue as well?

I know there is the info in introspection, so for example I think I can check if I _can_ generate the `lt` query, but if not and the query is of different "shape", how do I know what query to generate? Would I just return false from `isDateTimeFilter()` or something? A little bit confused here as to what to do.

sMteX commented:
@macrozone 

Ok, so I have made tons of changes today:
- I noticed a lot of filter objects are actually not supported yet, so I implemented those, hopefully (namely Int, String, Float, DateTime filters, along with their Nullable and Nested variants - they actually have the same structure but I guess they're used in different places so they have different names)
- majority of the comparators on these fields should now be implemented (`gt, gte, lt, lte, equals`, for strings also `contains, startsWith, endsWith`) - default is `equals`, except for Strings, where it's `contains`
- with the change of the regular expression for parsing the key, we can now process any kind of field names, including snake_case names (though I have heard a feedback, that using `:` as a delimiter for the comparer might strike the user as "this isn't part of the field name", like it's not common to have that in a name (e.g. even in programming languages, you can use `_` in variable names but not `:`) - `effectiveFrom_gt` vs `effectiveFrom:gt`)
- we now first check if the proposed special comparison query is **actually possible** by using `introspectionResults` (so e.g. in order for `effectiveFrom_gt` to filter "greater than" on the `effectiveFrom` field, the `effectiveFrom` field must exist, and must have the `gt` property on its type, otherwise it's processed as `field name = effectiveFrom_gt`
- I also added the FloatFilter to single values and array values as it was missing too

I tried adding tests for all of these scenarios but they seem very repetitive and might be a little too brittle (I also added a special Test entity to testing schema as it made more sense to not clutter existing models for this purpose).

I'm eager to hear your feedback on these changes :)

macrozone commented:
@sMteX that's exactly what i meant and it looks really nice.

I think the tests are good enough.

One thing: the project uses default semantic-release to create new releases, but i am not sure if it parsers `feature: xxxx` correctly as the default is `feat: xxxx`.

so can you change these commit-messages? you can also rebase them to one if you want.

sMteX commented:
@macrozone So the point is to have maybe multiple commits (or just 1), but with "feat: " prefix instead?

sMteX commented:
I suck at git 😅 I'll close this PR and open another one in a minute with just 1 commit.
